drivers/rtc/rtc_rpi_pico.c

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (c) 2024 Andrew Featherstone
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/rtc.h>
+#include <zephyr/irq.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/spinlock.h>
+
+#include <hardware/irq.h>
+#include <hardware/rtc.h>
+#include <hardware/regs/rtc.h>
+
+#include "rtc_utils.h"
+
+#define DT_DRV_COMPAT raspberrypi_pico_rtc
+
+#define CLK_DRV DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(0))
+#define CLK_ID  (clock_control_subsys_t) DT_INST_PHA_BY_IDX(0, clocks, 0, clk_id)
+
+/* struct tm start time:   1st, Jan, 1900 */
+#define TM_YEAR_REF 1900
+#ifdef CONFIG_RTC_ALARM
+static int rtc_rpi_pico_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				       struct rtc_time *timeptr);
+#endif
+struct rtc_rpi_pico_data {
+	struct k_spinlock lock;
+
+#ifdef CONFIG_RTC_ALARM
+	struct rtc_time alarm_time;
+	uint16_t alarm_mask;
+	rtc_alarm_callback alarm_callback;
+	void *alarm_user_data;
+	bool alarm_pending;
+#endif /* CONFIG_RTC_ALARM */
+};
+
+static struct rtc_rpi_pico_data rtc_data;
+
+LOG_MODULE_REGISTER(rtc_rpi, CONFIG_RTC_LOG_LEVEL);
+
+#ifdef CONFIG_RTC_ALARM
+static void rtc_rpi_isr(const struct device *dev)
+{
+	struct rtc_rpi_pico_data *data = dev->data;
+
+	rtc_alarm_callback callback;
+	void *user_data;
+
+	rtc_disable_alarm();
+
+	K_SPINLOCK(&data->lock) {
+		callback = data->alarm_callback;
+		user_data = data->alarm_user_data;
+	}
+
+	if (callback != NULL) {
+		callback(dev, 0, user_data);
+	} else {
+		data->alarm_pending = true;
+	}
+	/* re-enable the alarm. */
+	rtc_enable_alarm();
+}
+#endif
+
+static int rtc_rpi_pico_init(const struct device *dev)
+{
+	int ret;
+#ifdef CONFIG_RTC_ALARM
+	struct rtc_rpi_pico_data *data = dev->data;
+#endif
+
+	ret = clock_control_on(CLK_DRV, CLK_ID);
+	if (ret < 0) {
+		return ret;
+	}
+
+#ifdef CONFIG_RTC_ALARM
+	data->alarm_mask = 0;
+	data->alarm_callback = NULL;
+	data->alarm_pending = false;
+
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), rtc_rpi_isr, DEVICE_DT_INST_GET(0),
+		    0);
+	irq_enable(DT_INST_IRQN(0));
+#endif
+	rtc_init();
+	return 0;
+}
+
+static int rtc_rpi_pico_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	struct rtc_rpi_pico_data *data = dev->data;
+	int err = 0;
+
+	if (timeptr->tm_year + TM_YEAR_REF > 4095) {
```

Comment:
Please define this number.

Reply:
Done.

Reply:
@soburi It is expected for authors to resolve conversations if the change matches the request exactly :) I do get your point though :)